.changeset/slow-items-heal.md

```diff
@@ -0,0 +1,36 @@
+---
+"astro": minor
+---
+
+Adds experimental JSON Schema support for content collections of with `type: 'data'`.
+
+To enable this feature, add the experimental flag:
+
+```diff
+import { defineConfig } from 'astro/config';
+
+export default defineConfig({
+	experimental: {
++		contentCollectionJsonSchema: true
+	}
+});
+```
+
+To be able to use the schemas, you need to adapt your zod definition and reference it inside the json manually:
+
+```diff
+const test = defineCollection({
+	type: 'data',
+	schema: z.object({
++		"$schema": z.string().optional(),
```

Comment:
Ah this would be a nice limitation to avoid... Are you up to try to automatically allow the `$schema` property? It should probably _not_ be exposed in the resolved data, but even if it is, every data collection having `$schema?: string` sounds like a fair tradeoff.

Reply:
@natemoo-re I guess we should only add the `$schema` property if it is an object anyway, otherwise the user wouldn't have a way to set that anyway